File: libs/moqtail-rs/src/model/data/object.rs
// Copyright 2025 The MOQtail Authors
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

use std::fmt;

use bytes::Bytes;

use crate::model::common::location::Location;
use crate::model::error::ParseError;
use crate::model::extension_header::object_extension::ObjectExtension;

use super::constant::{ObjectForwardingPreference, ObjectStatus};
use super::datagram::Datagram;
use super::fetch_object::FetchObject;
use super::subgroup_object::SubgroupObject;

#[derive(Clone, PartialEq)]
pub struct Object {
  pub track_alias: u64,
  pub location: Location,
  pub publisher_priority: u8,
  pub forwarding_preference: ObjectForwardingPreference,
  pub subgroup_id: Option<u64>,
  pub status: ObjectStatus,
  pub extensions: Option<Vec<ObjectExtension>>,
  pub payload: Option<Bytes>,
}
impl fmt::Debug for Object {
  fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
    f.debug_struct("Object")
      .field("track_alias", &self.track_alias)
      .field("location", &self.location)
      .field("publisher_priority", &self.publisher_priority)
      .field("forwarding_preference", &self.forwarding_preference)
      .field("subgroup_id", &self.subgroup_id)
      .field("status", &self.status)
      .field("extensions", &self.extensions)
      .field("payload_length", &self.payload.as_ref().map(|p| p.len()))
      .finish()
  }
}

impl Object {
  /// Convert from a Datagram object.
  /// Returns (Object, end_of_group) since end_of_group is datagram-specific.
  /// When `publisher_priority` is None (DEFAULT_PRIORITY), `default_priority` is used.
  pub fn try_from_datagram(
    datagram: Datagram,
    default_priority: u8,
  ) -> Result<(Self, bool), ParseError> {
    let end_of_group = datagram.end_of_group;
    let priority = datagram.publisher_priority.unwrap_or(default_priority);

    let (status, payload) = if let Some(object_status) = datagram.object_status {
      (object_status, None)
    } else {
      (ObjectStatus::Normal, datagram.payload)
    };

    Ok((
      Object {
        track_alias: datagram.track_alias,
        location: Location {
          group: datagram.group_id,
          object: datagram.object_id,
        },
        publisher_priority: priority,
        forwarding_preference: ObjectForwardingPreference::Datagram,
        subgroup_id: None,
        status,
        extensions: datagram.extension_headers,
        payload,
      },
      end_of_group,
    ))
  }

  pub fn try_from_subgroup(
    subgroup_obj: SubgroupObject,
    track_alias: u64,               // Context from SubgroupHeader
    group_id: u64,                  // Context from SubgroupHeader
    subgroup_id: Option<u64>,       // Context from SubgroupHeader
    publisher_priority: Option<u8>, // Context from SubgroupHeader; None when using DEFAULT_PRIORITY
  ) -> Result<Self, ParseError> {
    Ok(Object {
      track_alias,
      location: Location {
        group: group_id,
        object: subgroup_obj.object_id,
      },
      publisher_priority: publisher_priority.unwrap_or(0), // Default to 0 if not specified
      forwarding_preference: ObjectForwardingPreference::Subgroup,
      subgroup_id,
      status: subgroup_obj.object_status.unwrap_or(ObjectStatus::Normal),
      extensions: subgroup_obj.extension_headers,
      payload: subgroup_obj.payload,
    })
  }

  pub fn try_from_fetch(fetch_obj: FetchObject, track_alias: u64) -> Result<Self, ParseError> {
    Ok(Object {
      track_alias, // Context from FetchHeader RequestId == Fetch Message Request Id, Fetch.track_alias
      location: Location {
        group: fetch_obj.group_id,
        object: fetch_obj.object_id,
      },
      publisher_priority: fetch_obj.publisher_priority,
      forwarding_preference: ObjectForwardingPreference::Subgroup,
      subgroup_id: Some(fetch_obj.subgroup_id),
      status: fetch_obj.object_status.unwrap_or(ObjectStatus::Normal),
      extensions: fetch_obj.extension_headers,
      payload: fetch_obj.payload,
    })
  }
}

impl Object {
  /// Convert to a Datagram object for wire transmission.
  ///
  /// # Arguments
  /// * `track_alias` - Track alias to use
  /// * `end_of_group` - Whether this is the last object in the group
  /// * `default_priority` - If provided and matches self.publisher_priority, omit priority on wire (DEFAULT_PRIORITY bit)
  pub fn try_into_datagram(
    self,
    track_alias: u64,
    end_of_group: bool,
    default_priority: Option<u8>,
  ) -> Result<Datagram, ParseError> {
    if self.forwarding_preference != ObjectForwardingPreference::Datagram {
      return Err(ParseError::CastingError {
        context: "Object::try_into_datagram(forwarding_preference)",
        from_type: "Object",
        to_type: "Datagram",
        details: "Forwarding preference must be Datagram".to_string(),
      });
    }

    // Determine whether to omit priority on wire
    let publisher_priority =
      if default_priority.is_some() && default_priority == Some(self.publisher_priority) {
        None
      } else {
        Some(self.publisher_priority)
      };

    if self.status != ObjectStatus::Normal {
      // Status datagram
      if self.payload.is_some() {
        return Err(ParseError::CastingError {
          context: "Object::try_into_datagram(payload)",
          from_type: "Object",
          to_type: "Datagram",
          details: "Payload must be None for non-Normal status".to_string(),
        });
      }
      Ok(Datagram::new_status(
        track_alias,
        self.location.group,
        self.location.object,
        publisher_priority,
        self.extensions,
        self.status,
      ))
    } else {
      // Payload datagram
      let payload = self.payload.ok_or(ParseError::CastingError {
        context: "Object::try_into_datagram(payload)",
        from_type: "Object",
        to_type: "Datagram",
        details: "Payload cannot be empty".to_string(),
      })?;
      Ok(Datagram::new_payload(
        track_alias,
        self.location.group,
        self.location.object,
        publisher_priority,
        self.extensions,
        payload,
        end_of_group,
      ))
    }
  }

  pub fn try_into_subgroup(self) -> Result<SubgroupObject, ParseError> {
    if self.forwarding_preference != ObjectForwardingPreference::Subgroup {
      return Err(ParseError::CastingError {
        context: "Object::try_into_subgroup(forwarding_preference)",
        from_type: "Object",
        to_type: "SubgroupObject",
        details: "Forwarding preference must be Subgroup".to_string(),
      });
    }
    let _subgroup_id = self.subgroup_id.ok_or(ParseError::CastingError {
      context: "Object::try_into_subgroup(subgroup_id)",
      from_type: "Object",
      to_type: "SubgroupObject",
      details: "Subgroup ID must be present for Subgroup forwarding".to_string(),
    })?;

    let (payload, object_status) = match self.status {
      ObjectStatus::Normal => (
        Some(self.payload.ok_or(ParseError::CastingError {
          context: "Object::try_into_subgroup(payload)",
          from_type: "Object",
          to_type: "SubgroupObject",
          details: "Payload must be present for Normal status".to_string(),
        })?),
        None,
      ),
      other_status => {
        if self.payload.is_some() {
          return Err(ParseError::CastingError {
            context: "Object::try_into_subgroup(payload)",
            from_type: "Object",
            to_type: "SubgroupObject",
            details: "Payload must be None for non-Normal status".to_string(),
          });
        }
        (None, Some(other_status))
      }
    };

    Ok(SubgroupObject {
      object_id: self.location.object,
      extension_headers: self.extensions,
      object_status,
      payload,
    })
  }

  pub fn try_into_fetch(self) -> Result<FetchObject, ParseError> {
    let (payload, object_status) = match self.status {
      ObjectStatus::Normal => (
        Some(self.payload.ok_or(ParseError::CastingError {
          context: "Object::try_into_fetch(payload)",
          from_type: "Object",
          to_type: "FetchObject",
          details: "Payload must be present for Normal status".to_string(),
        })?),
        None,
      ),
      other_status => {
        if self.payload.is_some() {
          return Err(ParseError::CastingError {
            context: "Object::try_into_fetch(payload)",
            from_type: "Object",
            to_type: "FetchObject",
            details: "Payload must be None for non-Normal status".to_string(),
          });
        }
        (None, Some(other_status))
      }
    };

    let subgroup_id = match self.forwarding_preference {
      ObjectForwardingPreference::Subgroup => self.subgroup_id.ok_or(ParseError::CastingError {
        context: "Object::try_into_fetch(subgroup_id)",
        from_type: "Object",
        to_type: "FetchObject",
        details: "Subgroup ID must be present for Subgroup forwarding".to_string(),
      })?,
      ObjectForwardingPreference::Datagram => {
        if self.subgroup_id.is_some() {
          return Err(ParseError::CastingError {
            context: "Object::try_into_fetch(subgroup_id)",
            from_type: "Object",
            to_type: "FetchObject",
            details: "Subgroup ID must not be present for Datagram forwarding".to_string(),
          });
        }
        self.location.object
      }
    };

    Ok(FetchObject {
      group_id: self.location.group,
      subgroup_id,
      object_id: self.location.object,
      publisher_priority: self.publisher_priority,
      extension_headers: self.extensions,
      payload,
      object_status,
    })
  }
}

File: libs/moqtail-ts/src/model/data/object.ts
/**
 * Copyright 2025 The MOQtail Authors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { KeyValuePair } from '../common/pair'
import { CastingError, ProtocolViolationError } from '../error/error'
import { ExtensionHeaders } from '../extension_header'
import { ObjectDatagramType, ObjectForwardingPreference, ObjectStatus, SubgroupHeaderType } from './constant'
import { Datagram } from './datagram'
import { FetchObject } from './fetch_object'
import { FullTrackName } from './full_track_name'
import { SubgroupObject } from './subgroup_object'
import { Location } from '../common/location'

export class MoqtObject {
  public readonly location: Location
  public readonly subgroupId: bigint | null

  private constructor(
    public readonly fullTrackName: FullTrackName,
    location: Location,
    public readonly publisherPriority: number,
    public readonly objectForwardingPreference: ObjectForwardingPreference,
    subgroupId: bigint | number | null,
    public readonly objectStatus: ObjectStatus,
    public readonly extensionHeaders: KeyValuePair[] | null,
    public readonly payload: Uint8Array | null,
  ) {
    this.location = location
    this.subgroupId = subgroupId !== null ? BigInt(subgroupId) : null
  }

  get groupId(): bigint {
    return this.location.group
  }
  get objectId(): bigint {
    return this.location.object
  }

  getSubgroupHeaderType(containsEnd: boolean, useDefaultPriority: boolean = false): SubgroupHeaderType {
    const hasExtensions = !!this.extensionHeaders && this.extensionHeaders.length > 0

    // Determine SUBGROUP_ID_MODE (bits 1-2):
    // 0b00 (0): Subgroup ID = 0 (absent from header)
    // 0b01 (1): Subgroup ID = first Object ID (absent from header)
    // 0b10 (2): Subgroup ID = explicit (present in header)
    let subgroupIdMode: 0 | 1 | 2
    if (this.subgroupId === null) {
      subgroupIdMode = 1 // first-object-ID mode
    } else if (this.subgroupId === 0n) {
      subgroupIdMode = 0 // zero mode
    } else {
      subgroupIdMode = 2 // explicit mode
    }

    return SubgroupHeaderType.fromProperties(hasExtensions, subgroupIdMode, containsEnd, useDefaultPriority)
  }

  isDatagram(): boolean {
    return this.objectForwardingPreference === ObjectForwardingPreference.Datagram
  }
  isSubgroup(): boolean {
    return this.objectForwardingPreference === ObjectForwardingPreference.Subgroup
  }
  isEndOfGroup(): boolean {
    return this.objectStatus === ObjectStatus.EndOfGroup
  }
  isEndOfTrack(): boolean {
    return this.objectStatus === ObjectStatus.EndOfTrack
  }
  doesNotExist(): boolean {
    return this.objectStatus === ObjectStatus.DoesNotExist
  }
  hasPayload(): boolean {
    return this.payload !== null
  }
  hasStatus(): boolean {
    return this.objectStatus !== ObjectStatus.Normal
  }

  static newWithPayload(
    fullTrackName: FullTrackName,
    location: Location,
    publisherPriority: number,
    objectForwardingPreference: ObjectForwardingPreference,
    subgroupId: bigint | number | null,
    extensionHeaders: KeyValuePair[] | null,
    payload: Uint8Array,
  ): MoqtObject {
    return new MoqtObject(
      fullTrackName,
      location,
      publisherPriority,
      objectForwardingPreference,
      subgroupId,
      ObjectStatus.Normal,
      extensionHeaders,
      payload,
    )
  }

  static newWithStatus(
    fullTrackName: FullTrackName,
    location: Location,
    publisherPriority: number,
    objectForwardingPreference: ObjectForwardingPreference,
    subgroupId: bigint | number | null,
    extensionHeaders: KeyValuePair[] | null,
    objectStatus: ObjectStatus,
  ): MoqtObject {
    return new MoqtObject(
      fullTrackName,
      location,
      publisherPriority,
      objectForwardingPreference,
      subgroupId,
      objectStatus,
      extensionHeaders,
      null,
    )
  }

  /**
   * Create a MoqtObject from a Datagram (Draft-16).
   * @param datagram - The received Datagram.
   * @param fullTrackName - The resolved full track name.
   * @param defaultPriority - The default publisher priority from the control message (used when DEFAULT_PRIORITY bit is set).
   */
  static fromDatagram(datagram: Datagram, fullTrackName: FullTrackName, defaultPriority: number = 0): MoqtObject {
    const priority = datagram.publisherPriority ?? defaultPriority
    if (ObjectDatagramType.isStatus(datagram.type)) {
      return new MoqtObject(
        fullTrackName,
        datagram.location,
        priority,
        ObjectForwardingPreference.Datagram,
        null,
        datagram.objectStatus!,
        datagram.extensionHeaders,
        null,
      )
    } else {
      return new MoqtObject(
        fullTrackName,
        datagram.location,
        priority,
        ObjectForwardingPreference.Datagram,
        null,
        ObjectStatus.Normal,
        datagram.extensionHeaders,
        datagram.payload,
      )
    }
  }

  /**
   * Returns the endOfGroup flag from the source Datagram.
   * This is separate from ObjectStatus.EndOfGroup - the flag indicates
   * this is the last object in the group even with Normal status.
   */
  static isDatagramEndOfGroup(datagram: Datagram): boolean {
    return datagram.endOfGroup
  }

  static fromFetchObject(fetchObject: FetchObject, fullTrackName: FullTrackName): MoqtObject {
    if (fetchObject.kind !== 'object') {
      throw new ProtocolViolationError(
        'MoqtObject.fromFetchObject',
        'EndOfRange markers cannot be converted to MoqtObject',
      )
    }
    const subgroupId =
      fetchObject.forwardingPreference === ObjectForwardingPreference.Subgroup ? fetchObject.subgroupId : null
    return new MoqtObject(
      fullTrackName,
      fetchObject.location,
      fetchObject.publisherPriority,
      fetchObject.forwardingPreference,
      subgroupId,
      ObjectStatus.Normal,
      fetchObject.extensionHeaders,
      fetchObject.payload && fetchObject.payload.length > 0 ? fetchObject.payload : null,
    )
  }

  static fromSubgroupObject(
    subgroupObject: SubgroupObject,
    groupId: bigint | number,
    publisherPriority: number | undefined,
    subgroupId: bigint | number | null,
    fullTrackName: FullTrackName,
  ): MoqtObject {
    return new MoqtObject(
      fullTrackName,
      new Location(groupId, subgroupObject.objectId),
      publisherPriority ?? 0,
      ObjectForwardingPreference.Subgroup,
      subgroupId,
      subgroupObject.objectStatus || ObjectStatus.Normal,
      subgroupObject.extensionHeaders,
      subgroupObject.payload,
    )
  }
  /**
   * Convert to Datagram for wire transmission (Draft-16).
   * Automatically creates a payload or status Datagram based on the object's state.
   * @param trackAlias - The track alias to use.
   * @param endOfGroup - Whether this is the last object in the group (only for payload datagrams).
   * @param defaultPriority - If provided and matches this object's priority, the DEFAULT_PRIORITY bit is set.
   */
  tryIntoDatagram(trackAlias: bigint | number, endOfGroup: boolean = false, defaultPriority?: number): Datagram {
    if (this.objectForwardingPreference !== ObjectForwardingPreference.Datagram) {
      throw new CastingError(
        'MoqtObject.tryIntoDatagram',
        'MoqtObject',
        'Datagram',
        'Object Forwarding Preference must be Datagram',
      )
    }

    const alias = BigInt(trackAlias)
    const priority =
      defaultPriority !== undefined && defaultPriority === this.publisherPriority ? null : this.publisherPriority

    if (this.hasStatus()) {
      return Datagram.newStatus(alias, this.groupId, this.objectId, priority, this.extensionHeaders, this.objectStatus)
    } else {
      if (!this.payload) {
        throw new ProtocolViolationError(
          'MoqtObject.tryIntoDatagram',
          'Object must have payload for payload Datagram conversion',
        )
      }
      return Datagram.newPayload(
        alias,
        this.groupId,
        this.objectId,
        priority,
        this.extensionHeaders,
        this.payload,
        endOfGroup,
      )
    }
  }
  tryIntoFetchObject(): FetchObject {
    // Draft-16 §10.4.4 FETCH objects carry no status; non-Normal status must be
    // surfaced as an EndOfRange marker by the caller instead.
    if (this.objectStatus !== ObjectStatus.Normal) {
      throw new CastingError(
        'MoqtObject.tryIntoFetchObject',
        'MoqtObject',
        'FetchObject',
        'non-Normal status cannot be serialized as a fetch payload; use EndOfRange instead',
      )
    }
    let subgroupId: bigint | number
    if (this.objectForwardingPreference === ObjectForwardingPreference.Subgroup) {
      if (this.subgroupId === null) {
        throw new ProtocolViolationError(
          'MoqtObject.tryIntoFetchObject',
          'Subgroup ID is required for Subgroup forwarding preference',
        )
      }
      subgroupId = this.subgroupId
    } else {
      if (this.subgroupId !== null) {
        throw new ProtocolViolationError(
          'MoqtObject.tryIntoFetchObject',
          'Subgroup ID must not be set for Datagram forwarding preference',
        )
      }
      subgroupId = this.objectId
    }
    return FetchObject.newObject(
      this.groupId,
      subgroupId,
      this.objectId,
      this.publisherPriority,
      this.objectForwardingPreference,
      this.extensionHeaders,
      this.payload ?? new Uint8Array(0),
    )
  }
  tryIntoSubgroupObject(): SubgroupObject {
    if (this.objectForwardingPreference !== ObjectForwardingPreference.Subgroup) {
      throw new CastingError(
        'MoqtObject.tryIntoSubgroupObject',
        'MoqtObject',
        'SubgroupObject',
        'Object Forwarding Preference must be Subgroup',
      )
    }
    if (this.objectStatus === ObjectStatus.Normal && this.payload) {
      return SubgroupObject.newWithPayload(this.location.object, this.extensionHeaders, this.payload)
    } else {
      return SubgroupObject.newWithStatus(this.location.object, this.extensionHeaders, this.objectStatus)
    }
  }
}

if (import.meta.vitest) {
  const { describe, test, expect } = import.meta.vitest

  describe('MoqtObject', () => {
    test('create object with payload', () => {
      const payload = new TextEncoder().encode('test payload')
      const extensionHeaders = new ExtensionHeaders().addAudioLevel(100).addCaptureTimestamp(0).build()
      const fullTrackName = FullTrackName.tryNew('test/demo', 'track1')
      const location = new Location(100n, 10n)
      const obj = MoqtObject.newWithPayload(
        fullTrackName,
        location,
        128,
        ObjectForwardingPreference.Subgroup,
        5n,
        extensionHeaders,
        payload,
      )

      expect(obj.location.equals(location)).toBe(true)
      expect(obj.groupId).toBe(100n)
      expect(obj.objectId).toBe(10n)
      expect(obj.publisherPriority).toBe(128)
      expect(obj.objectForwardingPreference).toBe(ObjectForwardingPreference.Subgroup)
      expect(obj.subgroupId).toBe(5n)
      expect(obj.objectStatus).toBe(ObjectStatus.Normal)
      expect(obj.extensionHeaders).toEqual(extensionHeaders)
      expect(obj.payload).toEqual(payload)
      expect(obj.hasPayload()).toBe(true)
      expect(obj.hasStatus()).toBe(false)
    })

    test('create object with status', () => {
      const fullTrackName = FullTrackName.tryNew('test/demo', 'track2')
      const location = new Location(200n, 20n)
      const obj = MoqtObject.newWithStatus(
        fullTrackName,
        location,
        64,
        ObjectForwardingPreference.Datagram,
        null,
        null,
        ObjectStatus.EndOfGroup,
      )

      expect(obj.location.equals(location)).toBe(true)
      expect(obj.groupId).toBe(200n)
      expect(obj.objectId).toBe(20n)
      expect(obj.publisherPriority).toBe(64)
      expect(obj.objectForwardingPreference).toBe(ObjectForwardingPreference.Datagram)
      expect(obj.subgroupId).toBe(null)
      expect(obj.objectStatus).toBe(ObjectStatus.EndOfGroup)
      expect(obj.extensionHeaders).toBe(null)
      expect(obj.payload).toBe(null)
      expect(obj.hasPayload()).toBe(false)
      expect(obj.hasStatus()).toBe(true)
    })

    test('convert from/to Datagram payload (Draft-16)', () => {
      const payload = new TextEncoder().encode('datagram payload')
      const fullTrackName = FullTrackName.tryNew('test/demo', 'track3')
      const datagram = Datagram.newPayload(42n, 100n, 10n, 128, null, payload, false)

      const moqtObj = MoqtObject.fromDatagram(datagram, fullTrackName)
      expect(moqtObj.objectForwardingPreference).toBe(ObjectForwardingPreference.Datagram)
      expect(moqtObj.subgroupId).toBe(null)
      expect(MoqtObject.isDatagramEndOfGroup(datagram)).toBe(false)

      const backToDatagram = moqtObj.tryIntoDatagram(42n, false)
      expect(backToDatagram.trackAlias).toBe(42n)
      expect(backToDatagram.groupId).toBe(100n)
      expect(backToDatagram.objectId).toBe(10n)
      expect(backToDatagram.endOfGroup).toBe(false)
      expect(backToDatagram.payload).toEqual(payload)
    })

    test('convert from/to Datagram with endOfGroup (Draft-16)', () => {
      const payload = new TextEncoder().encode('last in group')
      const fullTrackName = FullTrackName.tryNew('test/demo', 'track3')
      const datagram = Datagram.newPayload(42n, 100n, 10n, 128, null, payload, true)

      expect(MoqtObject.isDatagramEndOfGroup(datagram)).toBe(true)

      const moqtObj = MoqtObject.fromDatagram(datagram, fullTrackName)
      const backToDatagram = moqtObj.tryIntoDatagram(42n, true)
      expect(backToDatagram.endOfGroup).toBe(true)
    })

    test('convert from/to Datagram status (Draft-16)', () => {
      const fullTrackName = FullTrackName.tryNew('test/demo', 'track3')
      const datagram = Datagram.newStatus(42n, 100n, 10n, 128, null, ObjectStatus.EndOfGroup)

      const moqtObj = MoqtObject.fromDatagram(datagram, fullTrackName)
      expect(moqtObj.objectStatus).toBe(ObjectStatus.EndOfGroup)
      expect(moqtObj.payload).toBeNull()

      const backToDatagram = moqtObj.tryIntoDatagram(42n)
      expect(backToDatagram.objectStatus).toBe(ObjectStatus.EndOfGroup)
      expect(backToDatagram.payload).toBeNull()
    })

    test('convert Datagram with DEFAULT_PRIORITY (Draft-16)', () => {
      const payload = new TextEncoder().encode('default prio')
      const fullTrackName = FullTrackName.tryNew('test/demo', 'track3')
      const datagram = Datagram.newPayload(42n, 100n, 10n, null, null, payload, false)

      const moqtObj = MoqtObject.fromDatagram(datagram, fullTrackName, 64)
      expect(moqtObj.publisherPriority).toBe(64) // resolved from default

      const backToDatagram = moqtObj.tryIntoDatagram(42n, false, 64)
      expect(backToDatagram.publisherPriority).toBeNull() // matches default, so null
    })

    test('convert from/to FetchObject', () => {
      const payload = new TextEncoder().encode('fetch payload')
      const fullTrackName = FullTrackName.tryNew('test/demo', 'track4')
      const fetchObj = FetchObject.newObject(
        100n,
        5n,
        10n,
        128,
        ObjectForwardingPreference.Subgroup,
        null,
        payload,
      )

      const moqtObj = MoqtObject.fromFetchObject(fetchObj, fullTrackName)
      expect(moqtObj.objectForwardingPreference).toBe(ObjectForwardingPreference.Subgroup)
      expect(moqtObj.subgroupId).toBe(5n)
      expect(moqtObj.location.equals(fetchObj.location)).toBe(true)

      const backToFetch = moqtObj.tryIntoFetchObject()
      expect(backToFetch.groupId).toBe(100n)
      expect(backToFetch.subgroupId).toBe(5n)
      expect(backToFetch.objectId).toBe(10n)
      expect(backToFetch.payload).toEqual(payload)
    })

    test('utility functions', () => {
      const fullTrackName = FullTrackName.tryNew('test/demo', 'track5')
      const location1 = new Location(100n, 10n)
      const datagramObj = MoqtObject.newWithStatus(
        fullTrackName,
        location1,
        128,
        ObjectForwardingPreference.Datagram,
        null,
        null,
        ObjectStatus.EndOfTrack,
      )
      const location2 = new Location(100n, 10n)
      const subgroupObj = MoqtObject.newWithPayload(
        fullTrackName,
        location2,
        128,
        ObjectForwardingPreference.Subgroup,
        5n,
        null,
        new Uint8Array(),
      )

      expect(datagramObj.isDatagram()).toBe(true)
      expect(datagramObj.isSubgroup()).toBe(false)
      expect(subgroupObj.isDatagram()).toBe(false)
      expect(subgroupObj.isSubgroup()).toBe(true)

      expect(datagramObj.isEndOfTrack()).toBe(true)
      expect(datagramObj.isEndOfGroup()).toBe(false)
    })
  })
}
